app/src/main/java/com/limelight/Game.java

```diff
@@ -575,8 +575,13 @@ public boolean onKeyDown(int keyCode, KeyEvent event) {
         }
 
         boolean handled = false;
-        if (event.getDevice() == null ||
-                event.getDevice().getKeyboardType() != InputDevice.KEYBOARD_TYPE_ALPHABETIC) {
+
+        boolean detectedGamepad = event.getDevice() == null ? false :
+                ((event.getDevice().getSources() & InputDevice.SOURCE_JOYSTICK) != 0 ||
+                        (event.getDevice().getSources() & InputDevice.SOURCE_GAMEPAD) != 0);
+        if (detectedGamepad || (event.getDevice() == null ||
+                event.getDevice().getKeyboardType() != InputDevice.KEYBOARD_TYPE_ALPHABETIC
+        )) {
```

Comment:
The reason for the original logic is in the comment in the body of the if statement. Keyboards and gamepads have overlapping keycodes for dpad and directional arrows (KEYCODE_DPAD_*).

I chose the alphabetic check to try to reliably exclude gamepads from triggering that code path (which seems to have totally failed here). The Android framework considers a keyboard alphabetic if it has the Q key present, so I figured it should be a pretty safe way to easily distinguish what should be sent as a keyboard event vs a gamepad.

The extra DPAD_CENTER and BACK events are synthesized by the framework in response to not handling BUTTON_A and BUTTON_B.

I think this change is pretty safe. I was just curious why the existing code didn't work.